File: poker_design1/src/examples/PokerClient.java
/*
 * PokerClient.java
 *
 * George Ferguson, ferguson@cs.rochester.edu,  1 Oct 1998
 * Time-stamp: <Mon Oct 19 17:06:14 EDT 1998 ferguson>
 *
 * All this client does is hook up to the IPP server, echo lines received
 * from the server to stdout, prompt for replies from stdin, and send
 * the replies back to the server.
 *
 * Compile with: javac PokerClient.java
 * Run with: java PokerClient
 */

import java.io.*;
import java.net.*;

public class PokerClient extends Thread {
    // Members
    private BufferedReader in;
    private PrintWriter out;
    private static final String USAGE =
	"usage: java PokerClient [-server host[:port]]";
    // Constructor
    public PokerClient() {
    }
    // Methods
    public boolean connect(String host, int port) {
	try {
	    Socket sock = new Socket(host, port);
	    in = new BufferedReader(new InputStreamReader(sock.getInputStream()));
	    out = new PrintWriter(sock.getOutputStream(), true);
	    return true;
	} catch (IOException ex) {
	    System.err.println("IO error: " + ex);
	    return false;
	}
    }
    public void run() {
	BufferedReader stdin =
	    new BufferedReader(new InputStreamReader(System.in));
	PrintWriter stdout = new PrintWriter(System.out, true);
	try {
	    // Do forever...
	    while (true) {
		// Read msg from server
		String msg = in.readLine();
		// Copy to stdout
		stdout.println(msg);
		// EOF?
		if (msg == null) {
		    // Then exit
		    System.exit(0);
		} else if (msg.indexOf("?") > 0 || msg.startsWith("IPP")) {
		    // Otherwise if msg requires response, prompt for it
		    stdout.print("reply: ");
		    stdout.flush();
		    // Read response from stdin
		    String reply = stdin.readLine();
		    // And send it back to the server
		    out.println(reply);
		}
	    }
	} catch (IOException ex) {
	    System.err.println("IO error: " + ex);
	}
    }
    // Main for sample client
    public static void main(String[] argv) {
	// Defaults
	String host = "localhost";
	int  port = 9898;
	// Process command-line
	int argc = argv.length;
	int i;
	for (i=0; i < argc && argv[i].startsWith("-"); i++) {
	    if (argv[i].equals("-server")) {
		host = argv[++i];
		int j = host.indexOf(":");
		if (j > -1) {
		    port = new Integer(host.substring(j+1)).intValue();
		    host = host.substring(0, j);
		}
	    } else {
		System.err.println("unknown option: " + argv[i]);
		System.err.println(USAGE);
		System.exit(-1);
	    }
	}
	// Here we go
	PokerClient client = new PokerClient();
	if (client.connect(host, port)) {
	    client.run();
	}
    }
}

File: poker_design1/src/examples/pokerclient.c
/*
 * pokerclient.c
 *
 * George Ferguson, ferguson@cs.rochester.edu,  1 Oct 1998
 * Time-stamp: <Thu Oct  1 18:31:13 EDT 1998 ferguson>
 * All this client does is hook up to the IPP server, echo lines received
 * from the server to stdout, prompt for replies from stdin, and send
 * the replies back to the server.
 *
 * Compile with: gcc -o pokerclient pokerclient.c -lsocket -lnsl
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <netdb.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>

#define USAGE "usage: pokerclient [-server HOST:PORT]"

/*
 * Functions defined here:
 */
void main(int argc, char **argv);
static int connectToServer(char *host, int port);

/*
 * Data defined here:
 */
char *host = "localhost";
int port = 9898;

/*	-	-	-	-	-	-	-	-	*/

void
main(int argc, char **argv) {
    char *colon;
    int sock;
    FILE *in, *out;
    /* Parse command line */
    while (--argc > 0) {
	argv += 1;
	if (strcmp(argv[0], "-server") == 0) {
	    if (argc < 2) {
		fprintf(stderr, "missing argument for -server");
		exit(-1);
	    } else {
		host = argv[1];
		if ((colon=strchr(host, ':')) != NULL) {
		    sscanf(colon, ":%d", &port);
		    *colon = '\0';
		}
		argc -= 1;
		argv += 1;
	    }
	} else {
	    fprintf(stderr, "unknown option: %s", argv[0]);
	    fprintf(stderr, "%s\n", USAGE);
	}
    }
    /* Connect to server */
    printf("connecting to server %s:%d\n", host, port);
    if ((sock = connectToServer("localhost", 9898)) < 0) {
	fprintf(stderr, "couldn't connect to server %s:%d\n", host, port);
	exit(-1);
    }
    /* Create stdio streams */
    if ((in = fdopen(sock, "r")) == NULL) {
	perror("fdopen for read");
    }
    if ((out = fdopen(sock, "w")) == NULL) {
	perror("fdopen for write");
    }
    printf("connected to server %s:%d\n", host, port);
    /* Now do forever... */
    while (1) {
	char msg[256];
	/* Read a message from the server */
	if (fgets(msg, sizeof(msg), in) == NULL) {
	    break;
	}
	/* Copy it to stdout */
	printf("%s", msg);
	/* If msg requires response... */
	if (strchr(msg, '?') != NULL || strncmp(msg, "IPP", 3) == 0) {
	    /* ...Then prompt for reply */
	    printf("reply: ");
	    /* Read reply from stdin */
	    fgets(msg, sizeof(msg), stdin);
	    /* And send it on to the server */
	    fprintf(out, "%s", msg);
	    /* Flush socket connection */
	    fflush(out);
	}
    }
}

static int
connectToServer(char *host, int port)
{
    struct hostent *hent;
    u_long haddr;
    struct sockaddr_in saddr;
    int sockfd;

     /* Lookup address */
    if ((hent = gethostbyname(host)) == NULL) {
	perror("gethostbyname");
	return -1;
    }
    haddr = *((u_long*)(hent->h_addr_list[0]));
    /* Create socket */
    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
	perror("socket");
	return -1;
    }
    /* Connect to server */
    memset((char*)&saddr, '\0', sizeof(saddr));
    saddr.sin_family = htons((short)AF_INET);
    saddr.sin_port = htons((u_short)port);
    memcpy((char*)&(saddr.sin_addr),(char*)&haddr,sizeof(saddr.sin_addr));
    if (connect(sockfd,(struct sockaddr *)&saddr,sizeof(saddr)) < 0) {
	perror("connect");
	return -1;
    }
    /* Done */
    return sockfd;
}
